src/operator/output.rs

```diff
@@ -0,0 +1,311 @@
+/// Alternatively they can retrieve values from all mailboxes at once
+/// using [`take_from_all`](`OutputHandle::take_from_all`).
+/// If the stream carries relational data, the
+/// [`consolidate`](`OutputHandle::consolidate`) method can be used
+/// to combine output batches produced by all workes into a single
+/// batch.
+///
+/// Reading from a mailbox using any of these methods removes the value
+/// leaving the mailbox empty.  If the value is not read, it gets
+/// overwritten at the next clock cycle (i.e., during the next call to
+/// `step`).
+#[derive(Clone)]
+pub struct OutputHandle<T>(Arc<OutputHandleInternal<T>>);
+
+impl<T> OutputHandle<T>
+where
+    T: Send + Clone + 'static,
+{
+    fn new() -> Self {
+        match Runtime::runtime() {
+            None => Self(Arc::new(OutputHandleInternal::new(1))),
+            Some(runtime) => {
+                let output_id = runtime.sequence_next(Runtime::worker_index());
+
+                runtime
+                    .local_store()
+                    .entry(OutputId::new(output_id))
+                    .or_insert_with(|| {
+                        Self(Arc::new(OutputHandleInternal::new(runtime.num_workers())))
+                    })
+                    .value()
+                    .clone()
+            }
+        }
+    }
+
+    fn mailbox(&self, worker: usize) -> &Mailbox<Option<T>> {
+        self.0.mailbox(worker)
+    }
+
+    /// Read the value produced by `worker` worker thread during the last
+    /// clock cycle.
+    ///
+    /// This method is invoked between two subsequent
+    /// [`DBSPHandle::step`](`crate::DBSPHandle::step`)
+    /// calls to retrieve the value written to the stream during the last
+    /// clock cycle.  The first call is guaranteed to return a value
+    /// (since a synchronous circuit writes exactly one value to each
+    /// stream at every clock cycle).  It removes the value from the
+    /// mailbox, so subsequent calls will return `None`.
+    ///
+    /// Invoking this method in the middle of a clock cycle, i.e., during
+    /// the execution of `DBSPHandle::step`, may return either `None` or
+    /// `Some`, depending of whether a value has already been pushed to
+    /// the stream.
+    pub fn take_from_worker(&self, worker: usize) -> Option<T> {
+        self.0.take_from_worker(worker)
+    }
+
+    /// Read values produced by all worker threads during the last
+    /// clock cycle.
+    ///
+    /// This is equivalent to calling
+    /// [`take_from_worker`](`Self::take_from_worker`) for each
+    /// worker thread in order and storing all none-`None`
+    /// results in a vector.
+    pub fn take_from_all(&self) -> Vec<T> {
+        let num_workers = self.0.mailbox.len();
+        let mut res = Vec::with_capacity(num_workers);
+
+        for worker in 0..num_workers {
+            if let Some(v) = self.take_from_worker(worker) {
+                res.push(v);
+            }
+        }
+        res
```

Comment:
```suggestion

        let mut result = Vec::with_capacity(num_workers);
        for worker in 0..num_workers {
            if let Some(v) = self.take_from_worker(worker) {
                result.push(v);
            }
        }

        result
```

Reply:
Although honestly I kinda dislike the index loop we've got going on here, we could probably just get a slice/iterator of mailboxes to take from